src/types/redis_bitmap.cc

```diff
@@ -550,57 +550,6 @@ rocksdb::Status Bitmap::BitOp(BitOpFlags op_flag, const std::string &op_name, co
           memset(frag_res.get(), 0, frag_maxlen);
         }
 
-        /* Fast path: as far as we have data for all the input bitmaps we
-         * can take a fast path that performs much better than the
-         * vanilla algorithm. On ARM we skip the fast path since it will
-         * result in GCC compiling the code using multiple-words load/store
-         * operations that are not supported even in ARM >= v6. */
-#ifndef USE_ALIGNED_ACCESS
-        if (frag_minlen >= sizeof(uint64_t) * 4 && frag_numkeys <= 16) {
-          auto *lres = reinterpret_cast<uint64_t *>(frag_res.get());
-          const uint64_t *lp[16];
-          for (uint64_t i = 0; i < frag_numkeys; i++) {
-            lp[i] = reinterpret_cast<const uint64_t *>(fragments[i].data());
-          }
-          memcpy(frag_res.get(), fragments[0].data(), frag_minlen);
-          auto apply_fast_path_op = [&](auto op) {
-            // Note: kBitOpNot cannot use this op, it only applying
-            // to kBitOpAnd, kBitOpOr, kBitOpXor.
-            DCHECK(op_flag != kBitOpNot);
-            while (frag_minlen >= sizeof(uint64_t) * 4) {
-              for (uint64_t i = 1; i < frag_numkeys; i++) {
-                op(lres[0], lp[i][0]);
-                op(lres[1], lp[i][1]);
-                op(lres[2], lp[i][2]);
-                op(lres[3], lp[i][3]);
-                lp[i] += 4;
-              }
-              lres += 4;
-              j += sizeof(uint64_t) * 4;
-              frag_minlen -= sizeof(uint64_t) * 4;
-            }
-          };
-
-          if (op_flag == kBitOpAnd) {
-            apply_fast_path_op([](uint64_t &a, uint64_t b) { a &= b; });
-          } else if (op_flag == kBitOpOr) {
-            apply_fast_path_op([](uint64_t &a, uint64_t b) { a |= b; });
-          } else if (op_flag == kBitOpXor) {
-            apply_fast_path_op([](uint64_t &a, uint64_t b) { a ^= b; });
-          } else if (op_flag == kBitOpNot) {
-            while (frag_minlen >= sizeof(uint64_t) * 4) {
-              lres[0] = ~lres[0];
-              lres[1] = ~lres[1];
-              lres[2] = ~lres[2];
-              lres[3] = ~lres[3];
-              lres += 4;
-              j += sizeof(uint64_t) * 4;
-              frag_minlen -= sizeof(uint64_t) * 4;
-            }
-          }
-        }
-#endif
```

Comment:
I'm wondering if the performance difference between this fast path and vanilla algorithm is large.

Reply:
Personally I think https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/bitmap_ops.h#L160-L242 would be better, I'll do a benchmark there

Reply:
I've checked the current impl, and I guess the fast path would be faster😅

Reply:
FYI: https://godbolt.org/z/dnK6xbzvj
@PragmaTwice @git-hulk

Reply:
You can use https://quick-bench.com/ which can generate a chart.

Reply:
I revert the part because the underlying code is too slow 😅

Besides, I change to use memcpy to generalize it. I'll try a benchmark later

Reply:
https://github.com/apache/arrow/commit/76cebfa88a6bffc8e1422be23fa18f44806cd245
Lucky, arrow has similiar testing here. On My MacOS M1Pro with Release -O2: